src/NLog/Contexts/GlobalContext.cs

```diff
@@ -0,0 +1,220 @@
+        {
+            get
+            {
+                return this.TryGet(key);
+            }
+            set
+            {
+                lock (syncRoot)
+                {
+                    this.dict[key] = value;
+                    if (!keys.Contains(key))
+                        keys.Add(key);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Clears the content.
+        /// </summary>
+        public void Clear()
+        {
+            lock (syncRoot)
+            {
+                this.dict.Clear();
+                this.keys.Clear();
+            }
+        }
+
+        /// <summary>
+        /// Checks whether the specified key exists in the Context.
+        /// </summary>
+        /// <param name="key">key name.</param>
+        /// <returns>A boolean indicating whether the specified key exists in current context.</returns>
+        public bool Contains(string key)
+        {
+            lock (syncRoot)
+            {
+                return this.dict.ContainsKey(key);
+            }
+        }
+
+        /// <summary>
+        /// Removes the specified key from the Context.
+        /// </summary>
+        /// <param name="key">key name.</param>
+        public void Remove(string key)
+        {
+            lock (syncRoot)
+            {
+                this.dict.Remove(key);
+            }
+        }
+
+        /// <summary>
+        /// Sets an key in the context.
+        /// </summary>
+        /// <param name="key"></param>
+        /// <param name="value"></param>
+        public void Set(string key, object value)
+        {
+            this[key] = value;
+        }
+
+        /// <summary>
+        /// Gets the Global Context named item.
+        /// </summary>
+        /// <param name="item">Item name.</param>
+        /// <returns>The item value, if defined; otherwise <c>null</c>.</returns>
+        private object TryGet(string item)
+        {
+            lock (syncRoot)
+            {
+                object o;
+                if (!this.dict.TryGetValue(item, out o))
+                    o = null;
+
+                return o;
+            }
+        }
+
+        /// <summary>
+        /// Gets the number of items in the collection.
+        /// </summary>
+        public int Count
+        {
+            get
+            {
+                lock (syncRoot)
+                {
+                    return this.dict.Count;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets the Global Diagnostics Context item.
+        /// </summary>
+        /// <param name="key">Item name.</param>
+        /// <param name="formatProvider"><see cref="IFormatProvider"/> to use when converting the item's value to a string.</param>
+        /// <returns>The value of <paramref name="key"/> as a string, if defined; otherwise <see cref="String.Empty"/>.</returns>
+        public string GetFormatted(string key, IFormatProvider formatProvider)
```

Comment:
This will needed if some using the API wants to get the formatted string for some reason.